Index: gnome-keyring/daemon/gkd-main.c
===================================================================
--- gnome-keyring.orig/daemon/gkd-main.c
+++ gnome-keyring/daemon/gkd-main.c
@@ -574,8 +574,14 @@ discover_other_daemon (DiscoverFunc call
 	/* An environment variable from an already running daemon */
 	control_env = g_getenv (GKD_UTIL_ENV_CONTROL);
 	if (control_env && control_env[0]) {
-		if ((callback)(control_env))
-			return TRUE;
+		struct stat st;
+		if (lstat(control_env, &st) < 0) {
+			g_warning("discover other daemon: couldn't access gnome keyring socket: %s: %s", control_env, strerror(errno));
+			control_env = NULL;
+		} else {
+			if ((callback)(control_env))
+				return TRUE;
+		}
 	}
 
 	/* Or the default location when no evironment variable */
Index: gnome-keyring/pam/gkr-pam-module.c
===================================================================
--- gnome-keyring.orig/pam/gkr-pam-module.c
+++ gnome-keyring/pam/gkr-pam-module.c
@@ -1052,7 +1052,7 @@ pam_chauthtok_update (pam_handle_t *ph,
 	 * Likely the daemon is being started later in the session if we weren't
 	 * allowed to autostart it here. Store the password for our session handler
 	 */
-	if (!(args & ARG_AUTO_START))
+	if (!(args & ARG_AUTO_START) && ret == PAM_SUCCESS)
 		stash_password_for_session (ph, password);
 
 	return ret;
